File: trees/#102_Binary_Tree_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q = deque()
        if root:
            q.append(root)
        res = []
        while(len(q) > 0):
            level_len = len(q)
            level = []
            for i in range(level_len):
                curr = q.popleft()
                level.append(curr.val)
                print(f"{curr.val} ", end="")
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            res.append(level)
            print()
        return res

File: Leetcode/Cpp/#102_Binary_Tree Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        queue<TreeNode*> q;
        q.push(root);
        while(q.size() > 0) {
            int level_size = q.size();
            vector<int> level;
            for(int i = 0; i < level_size; i++) {
                TreeNode* curr = q.front();
                q.pop();
                if(curr) {
                    level.push_back(curr->val);
                    cout << curr->val << " ";
                    q.push(curr->left);
                    q.push(curr->right);
                }
            }
            if(!level.empty()) {
                res.push_back(level);
            }
            cout << endl;
        }
        return res;
    }
};
